Network chapter (#227)
hadley/ggplot2-book: networks.Rmd

# Networks {#networks}

```{r setup, include = FALSE}
source("common.R")
columns(1, 2 / 3)
set.seed(1)
```

> Like maps and spatial data, networks and graphs occupy a special part of the visualization landscape, but whereas spatial data mostly differ from regular plotting in their use of projections, networks bring their own data structure as well as their own visualization paradigms to the table. Because of these complications networks are not directly supported in ggplot2. Several efforts over the years have tried to add this missing piece and in this chapter we will see how to use ggraph for network visualization. Other packages that offers some of the same functionality includes geomnet, ggnetwork, and GGally for regular network plots, and ggtree and ggdendro for tree visualization specifically.

Reviewer: Can you please add links to the package names?

Author: Which links should we choose? Github, CRAN, pkgdown sites if any? Considering recent misfortunes with linking to domains not in the hands of companies I'd be weary of doing the third, though that is usually the most useful link

Reviewer: pkgdown is safe I think. I think the risk of domain highjacking is low for most sites.

```{r}
graph %>% 
  activate(nodes) %>% 
  mutate(centrality = centrality_pagerank()) %>% 
  arrange(desc(centrality))
```

### Want more?

This is just a brief glimpse into tidygraph, for the sake of understanding ggraph. If you are interested in learning more, the tidygraph website gives an overview of all the functionalities in the package: <https://tidygraph.data-imaginist.com>

## Visualizing networks

> ggraph builds on top of tidygraph and ggplot2 to allow a complete and familiar grammar of graphics for network data. Still, it is a little different from most ggplot2 extension packages since it works with another data type that is fundamentally different from tabular data. More so, most network visualizations don't concern themselves with mapping variables to `x` and `y` aesthetics since they are concerned with shoving the network topology more than relations between two variables. In order to show network topology the concept of layouts are employed. Layouts are algorithms that use the network structure to calculate (often arbitrary) `x` and `y` values for each node that can then be used for visualization purpose.

Reviewer: I think I'd emphasise that with tabular data `x` and `y` are almost always mapped to existing variables, but with network data, `x` and `y` are almost always generated from the network topology.

### Setting up the visualization

Whereas a normal ggplot2 plot is initialized with a `ggplot()` call, a ggraph plot is initialized with a `ggraph()` call. The first argument is the data, which can be a tbl\_graph or any object convertible to one. The second argument is a layout function and any further argument will be passed on to that function. The default layout will choose an appropriate layout based on the type of graph you provide, but while it is often a descent starting point you should always take control and explore the different layouts available --- networks are nefarious for their ability to show non-existing or exaggerated relations in some layouts.

#### Specifying a layout

> The layout argument can either take a string or a function. If a string is provided, the name will be matched to one of the build in layouts (of which there are many). If a function is provided it is assumed that the function takes a tbl\_graph and returns a data frame with at least an x and y column and with the same number of rows as there are nodes in the input graph. Below we can see examples of using the default layout, specifying a specific layout, and providing a layout function directly:

Reviewer: I think you can leave the use of a function for the docs, and instead show another existing layout.

```{r}
library(ggraph)
ggraph(hs_graph) + 
  geom_edge_link() + 
  geom_node_point()
```

```{r}
ggraph(hs_graph, layout = "drl") + 
  geom_edge_link() + 
  geom_node_point()
```

```{r}
random_layout <- function(x) {
  n_nodes <- with_graph(x, graph_order())
  data.frame(
    x = runif(n_nodes),
    y = runif(n_nodes)
  )
}
ggraph(hs_graph, layout = random_layout) + 
  geom_edge_link() + 
  geom_node_point()
```

> In order to show the graph above we're using the `geom_edge_link()` and `geom_node_point()` functions, and while we have not yet discussed these the do exactly what you may imagine: drawing nodes as points and edges as straight lines.

Reviewer: Include link to list of layout algorithms?

#### Circularity

> Some layouts may be used in both a linear and circular version. The correct way to change this in ggplot2 would be to use `coord_polar()` to change the coordinate system, but since we only want to change the position of nodes in the layout, and not affect the geoms in any other way, this is a function of the layout. The following can show the difference:

Reviewer: ```suggestion
Some layouts may be used in both a linear and circular version. The correct way to change this in ggplot2 would be to use `coord_polar()` to change the coordinate system, but since we only want to change the position of nodes in the layout, and not affect the edges, this is a function of the layout. The following can show the difference:
```

```{r}
ggraph(luv_graph, layout = 'dendrogram', circular = TRUE) + 
  geom_edge_link() + 
  coord_fixed()
```

```{r}
ggraph(luv_graph, layout = 'dendrogram') + 
  geom_edge_link() + 
  coord_polar() + 
  scale_y_reverse()
```

As we can see, using `coord_polar()` will bend our edges, hardly a desirable choice.

> There are more to layouts than described here. The Getting Started guide to layouts will tell you even more: <https://ggraph.data-imaginist.com/articles/Layouts.html>

Reviewer: Oh, just move this link up higher

Author: Should I lead with this? I think it makes sense to have it at the end? Same for edges and nodes...

Reviewer: Yeah, I think introduce the links in the first paragraphs.

There are more ways to draw edges than simple straight lines. Some are specific to trees or specific layouts, but many are general purpose. One specific use-case for another edge type is when you have multiple edges running between the same nodes. Drawing them as straight lines will obscure the multiplicity of the edges, which is e.g. apparent with the highschool graph where multiple parallel edges are present but invisible in the plots above. In order to show parallel edges you can either use `geom_edge_fan()` or `geom_edge_parallel()`:

```{r}
ggraph(hs_graph, layout = "stress") + 
  geom_edge_fan()
```

```{r}
ggraph(hs_graph, layout = "stress") + 
  geom_edge_parallel()
```

It is clear that these geoms should only be used for relatively simple graphs since they increase the amount of clutter and overplotting in the plot. Looking at trees and specifically dendrograms, one commonly used edge type is the elbow edge:

```{r}
ggraph(luv_graph, layout = "dendrogram", height = height) + 
  geom_edge_elbow()
```

`geom_edge_bend()` and `geom_edge_diagonal()` are smoother versions of this.

> 

Reviewer: Link to list of edge geoms?
